R/marginal_model.R

```diff
@@ -1,41 +1,61 @@
 #' Prepare marginal model to pass through to `brms`
 #'
 #' @param data A `data.frame` containing line list data
+#' @param ... Additional arguments passed to methods.
 #' @family marginal_model
 #' @export
-as_epidist_marginal_model <- function(data) {
+as_epidist_marginal_model <- function(data, ...) {
   UseMethod("as_epidist_marginal_model")
 }
 
 #' The marginal model method for `epidist_linelist_data` objects
 #'
 #' @param data An `epidist_linelist_data` object
+#' @param obs_time_threshold Ratio used to determine threshold for setting
+#'   relative observation times to Inf. Observation times greater than
+#'   `obs_time_threshold` times the maximum delay will be set to Inf to improve
+#'   model efficiency by reducing the number of unique observation times.
+#'   Default is 2.
+#' @param ... Not used in this method.
 #' @method as_epidist_marginal_model epidist_linelist_data
 #' @family marginal_model
 #' @autoglobal
 #' @export
-as_epidist_marginal_model.epidist_linelist_data <- function(data) {
+as_epidist_marginal_model.epidist_linelist_data <- function(
+    data, obs_time_threshold = 2, ...) {
   assert_epidist(data)
 
-  # Here we do the processing to turn an epidist_linelist_data into an aggregate
-  # dataset. In the future this would be refactored into a function which
-  # converts from linelist data to aggregate data, and a function which goes
-  # from aggregate data into the marginal model class
   data <- data |>
     mutate(
-      pwindow = ifelse(
-        .data$stime_lwr < .data$ptime_upr,
-        .data$stime_upr - .data$ptime_lwr,
-        .data$ptime_upr - .data$ptime_lwr
-      ),
-      delay = .data$stime_lwr - .data$ptime_lwr
-    ) |>
-    dplyr::group_by(delay) |>
-    dplyr::summarise(
-      count = dplyr::n(),
-      pwindow = ifelse(all(pwindow == pwindow[1]), pwindow[1], NA)
+      pwindow = .data$ptime_upr - .data$ptime_lwr,
+      swindow = .data$stime_upr - .data$stime_lwr,
+      relative_obs_time = .data$obs_time - .data$ptime_lwr,
+      orig_relative_obs_time = .data$obs_time - .data$ptime_lwr,
+      delay_lwr = .data$stime_lwr - .data$ptime_lwr,
+      delay_upr = .data$stime_upr - .data$ptime_lwr,
+      n = 1
     )
 
+  # Calculate maximum delay
+  max_delay <- max(data$delay_upr, na.rm = TRUE)
+  threshold <- max_delay * obs_time_threshold
+
+  # Count observations beyond threshold
+  n_beyond <- sum(data$relative_obs_time > threshold, na.rm = TRUE)
+
+  if (n_beyond > 0) {
+    cli::cli_inform(c(
+      "!" = paste0(
+        "Setting {n_beyond} observation time{?s} beyond ",
```

Comment:
More generally seems like a methods question here about batching together observations which are sufficiently similar. Do we have existing issue for that?

Reply:
no as you can only do so in the marginal model

Reply:
I don't think it only applying to the marginal model is sufficient reason not to consider it